### PR TITLE
feat: Added checking of duplicate struct fields and associated tests to weeder

### PR DESCRIPTION
**Pull request by dghosef**
This is the first of the well-formedness checks listed in #7938.

Would welcome any suggestions for error messages(or approval if you think they are ok) - I tried to copy similar error messages.

### CODE REVIEW

**Comment by magnus-madsen** [comment]
Can you post a few screenshots of the error messages?

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 1]
```diff
@@ -460,7 +460,12 @@ object Weeder2 {
         traverse(fields)(visitStructField)
       ) {
         (doc, ann, mods, ident, tparams, fields) =>
-        // TODO: Validation, e.g., never duplicate names
+        // Ensure that each name is unique
+        val errors = getDuplicates(fields, (f: StructField) => f.ident.name)
+        if (errors.nonEmpty) {
+          Validation.toHardFailure(DuplicateStructField(ident.loc, errors.head._1.ident.name, ident.name, errors.head._1.ident.loc,
```
[the remark this answers is not in the document]
Comment (dghosef) [reply main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:466]: > We should discuss how to make this a soft failure.
> 
> My suggestion: Drop all duplicate fields after the first. Can you try that?

So
```
struct {
  a: Int32,
  a: Float32
  b: Int32
}
```
will just become
```
struct {
  a: Int32,
  b: Int32
}
```
?

**Comment by dghosef** [comment]
> Can you post a few screenshots of the error messages?

Yea one sec I just realized a bug

**Comment by dghosef** [comment]
For the following code:

```
struct Bad {
    a: Int32,
    a: Float32
}
```

The error message is:
```
-- Syntax Error -------------------------------------------------- /Users/dghosef/flix/main/test/flix/Test.Dec.Struct.flix

>> Multiple declarations of field 'a' in struct 'Bad'

36 | struct Bad {
            ^^^
            struct declaration has duplicate fields

37 |     a: Int32,
         ^
         the first occurrence was here

38 |     a: Float32
         ^
         the second occurrence was here
```

**Comment by magnus-madsen** [comment]
@dghosef You can mark comments as "resolved" when you fix them.

Unfortunately, we cannot merge in the tests yet because it makes the CI fail. I suggest to either mark them as ignored or move them to your test PR.

**Comment by dghosef** [comment]
> @dghosef You can mark comments as "resolved" when you fix them.
> 
> Unfortunately, we cannot merge in the tests yet because it makes the CI fail. I suggest to either mark them as ignored or move them to your test PR.

Sounds good - marked as ignored for now

File: main/src/ca/uwaterloo/flix/language/errors/WeederError.scala [hunk 2]
```diff
@@ -143,6 +143,37 @@ object WeederError {
...
+  /**
+    * An error raised to indicate a struct contains duplicate fields
+    *
+    * @param loc the location of the struct declaration
+    * @param fieldName the name of the field
+    * @param structName the name of the struct
+    * @param field1 the location of the first field
+    * @param field2 the location of the second field
+    */
+  case class DuplicateStructField(loc: SourceLocation, fieldName: String, structName: String, field1: SourceLocation, field2: SourceLocation) extends WeederError with Recoverable {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/errors/WeederError.scala:155]: I would like the following order of fields:

structName
fieldName
field1Loc
field2Loc
loc

This matches how we do it in most other error messages.

Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/errors/WeederError.scala:155]: I think this file is alphabetically sorted, so this. error class should be moved up a bit.

File: main/src/ca/uwaterloo/flix/language/errors/WeederError.scala [hunk 3]
```diff
@@ -143,6 +143,37 @@ object WeederError {
...
+    def summary: String = s"Multiple declarations of field '$fieldName' in struct '$structName'"
+
+    def message(formatter: Formatter): String = {
+      import formatter._
+      s""">> Multiple declarations of field '$fieldName' in struct '$structName'
+         |
+         |${code(loc, "struct declaration has duplicate fields")}
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/errors/WeederError.scala:162]: I think "struct has duplicate fields" is good enough :)

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 4]
```diff
@@ -460,10 +460,20 @@ object Weeder2 {
...
         (doc, ann, mods, ident, tparams, fields) =>
-        // TODO: Validation, e.g., never duplicate names
-        Validation.success(Declaration.Struct(
-          doc, ann, mods, ident, tparams, fields.sortBy(_.loc), tree.loc
+        // Ensure that each name is unique
+        val errors = getDuplicates(fields, (f: StructField) => f.ident.name)
+        // For each field, only keep the first occurrence of the name
+        val groupedByName = fields.groupBy(_.ident.name)
+        val filteredFields = groupedByName.values.map(_.head).toList
+        val result = Validation.success(Declaration.Struct(
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:468]: You can inline this local result variable. That way there is no confusion.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:468]: In fact, its probably more clean todo:

if (errors.isEmpty) 
 ...
else 
 ...

so you handle the positive case first.

File: main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala [hunk 5]
```diff
@@ -460,10 +460,23 @@ object Weeder2 {
...
-        // TODO: Validation, e.g., never duplicate names
-        Validation.success(Declaration.Struct(
-          doc, ann, mods, ident, tparams, fields.sortBy(_.loc), tree.loc
-        ))
+        // Ensure that each name is unique
+        val errors = getDuplicates(fields, (f: StructField) => f.ident.name)
+        // For each field, only keep the first occurrence of the name
+        val groupedByName = fields.groupBy(_.ident.name)
+        val filteredFields = groupedByName.values.map(_.head).toList
+        if (errors.isEmpty) {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:468]: I just remembered we can use Validation.toSuccessOrSoftFailure

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:468]: Ah, but that won't work if we have to drop extra fields.

**Comment by magnus-madsen** [comment]
Merged! - when CI completes. Thanks!

**Pull request merged**